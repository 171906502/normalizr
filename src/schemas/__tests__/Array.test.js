/* eslint-env jest */
import { normalize, schema } from '../../';

describe(schema.Array.name, () => {
  describe('simple Arrays', () => {
    it('does a thing', () => {
      it(`normalizes plain arrays as shorthand for ${schema.Array.name}`, () => {
        const userSchema = new schema.Entity('user');
        expect(normalize([ { id: 1 }, { id: 2 } ], [ userSchema ])).toMatchSnapshot();
      });
    });

    it('throws an error if created with more than one schema', () => {
      const userSchema = new schema.Entity('users');
      const catSchema = new schema.Entity('cats');
      expect(() => normalize([ { id: 1 } ], [ catSchema, userSchema ])).toThrow();
    });

    it('passes its parent to its children when normalizing', () => {
      const processStrategy = (entity, parent, key) => {
        return { ...entity, parentId: parent.id, parentKey: key };
      };
      const childEntity = new schema.Entity('children', {}, { processStrategy });
      const parentEntity = new schema.Entity('parents', {
        children: [ childEntity ]
      });

      expect(normalize({
        id: 1, content: 'parent', children: [ { id: 4, content: 'child' } ]
      }, parentEntity)).toMatchSnapshot();
    });
  });

  describe('multiple schema', () => {
    it('normalizes multiple entities', () => {
      const inferSchemaFn = jest.fn((input, parent, key) => input.type || 'dogs');
      const userSchema = new schema.Entity('users');
      const catSchema = new schema.Entity('cats');
      const listSchema = new schema.Array({
        users: userSchema,
        cats: catSchema,
        dogs: {}
      }, inferSchemaFn);

      expect(normalize([
        { type: 'cats', id: '123' },
        { type: 'users', id: '123' },
        { type: 'cats', id: '456' },
        { id: '789' }
      ], listSchema)).toMatchSnapshot();
      expect(inferSchemaFn.mock.calls).toMatchSnapshot();
    });
  });
});
