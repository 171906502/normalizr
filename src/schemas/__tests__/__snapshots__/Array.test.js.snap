exports[`ArraySchema multiple schema normalizes multiple entities 1`] = `
Object {
  "entities": Object {
    "cats": Object {
      "123": Object {
        "id": "123",
        "type": "cats",
      },
      "456": Object {
        "id": "456",
        "type": "cats",
      },
    },
    "users": Object {
      "123": Object {
        "id": "123",
        "type": "users",
      },
    },
  },
  "result": Array [
    "123",
    "123",
    "456",
    Object {
      "id": "789",
    },
  ],
}
`;

exports[`ArraySchema multiple schema normalizes multiple entities 2`] = `
Array [
  Array [
    Object {
      "id": "123",
      "type": "cats",
    },
    Array [
      Object {
        "id": "123",
        "type": "cats",
      },
      Object {
        "id": "123",
        "type": "users",
      },
      Object {
        "id": "456",
        "type": "cats",
      },
      Object {
        "id": "789",
      },
    ],
    0,
  ],
  Array [
    Object {
      "id": "123",
      "type": "users",
    },
    Array [
      Object {
        "id": "123",
        "type": "cats",
      },
      Object {
        "id": "123",
        "type": "users",
      },
      Object {
        "id": "456",
        "type": "cats",
      },
      Object {
        "id": "789",
      },
    ],
    1,
  ],
  Array [
    Object {
      "id": "456",
      "type": "cats",
    },
    Array [
      Object {
        "id": "123",
        "type": "cats",
      },
      Object {
        "id": "123",
        "type": "users",
      },
      Object {
        "id": "456",
        "type": "cats",
      },
      Object {
        "id": "789",
      },
    ],
    2,
  ],
  Array [
    Object {
      "id": "789",
    },
    Array [
      Object {
        "id": "123",
        "type": "cats",
      },
      Object {
        "id": "123",
        "type": "users",
      },
      Object {
        "id": "456",
        "type": "cats",
      },
      Object {
        "id": "789",
      },
    ],
    3,
  ],
]
`;

exports[`ArraySchema simple Arrays passes its parent to its children when normalizing 1`] = `
Object {
  "entities": Object {
    "children": Object {
      "4": Object {
        "content": "child",
        "id": 4,
        "parentId": 1,
        "parentKey": "children",
      },
    },
    "parents": Object {
      "1": Object {
        "children": Array [
          4,
        ],
        "content": "parent",
        "id": 1,
      },
    },
  },
  "result": 1,
}
`;
